frame/contracts/src/migration.rs

```diff
@@ -15,459 +15,556 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use crate::{BalanceOf, CodeHash, Config, Pallet, TrieId, Weight};
-use codec::{Decode, Encode};
+pub mod v10;
+pub mod v11;
+pub mod v9;
+
+use crate::{weights::WeightInfo, Config, Error, MigrationInProgress, Pallet, Weight, LOG_TARGET};
+use codec::{Codec, Decode};
 use frame_support::{
 	codec,
 	pallet_prelude::*,
-	storage::migration,
-	storage_alias,
-	traits::{Get, OnRuntimeUpgrade},
-	Identity, Twox64Concat,
+	traits::{ConstU32, OnRuntimeUpgrade},
 };
-use sp_runtime::traits::Saturating;
-use sp_std::{marker::PhantomData, prelude::*};
+use sp_std::marker::PhantomData;
 
-/// Performs all necessary migrations based on `StorageVersion`.
-pub struct Migration<T: Config>(PhantomData<T>);
-impl<T: Config> OnRuntimeUpgrade for Migration<T> {
-	fn on_runtime_upgrade() -> Weight {
-		let version = <Pallet<T>>::on_chain_storage_version();
-		let mut weight = Weight::zero();
+#[cfg(feature = "try-runtime")]
+use sp_std::prelude::*;
 
-		if version < 4 {
-			v4::migrate::<T>(&mut weight);
-		}
+const PROOF_ENCODE: &str = "Tuple::max_encoded_len() < Cursor::max_encoded_len()` is verified in `Self::integrity_test()`; qed";
+const PROOF_DECODE: &str =
+	"We encode to the same type in this trait only. No other code touches this item; qed";
 
-		if version < 5 {
-			v5::migrate::<T>(&mut weight);
-		}
+fn invalid_version(version: StorageVersion) -> ! {
+	panic!("Required migration {version:?} not supported by this runtime. This is a bug.");
+}
 
-		if version < 6 {
-			v6::migrate::<T>(&mut weight);
-		}
+pub type Cursor = BoundedVec<u8, ConstU32<1024>>;
 
-		if version < 7 {
-			v7::migrate::<T>(&mut weight);
-		}
+// In benchmark and tests we use noop migrations, to test and bench the migration framework itself.
+#[cfg(not(any(feature = "runtime-benchmarks", test)))]
+type Migrations<T> = (v9::Migration<T>, v10::Migration<T>, v11::Migration<T>);
 
-		if version < 8 {
-			v8::migrate::<T>(&mut weight);
-		}
+/// IsFinished describes whether a migration is finished or not.
+pub enum IsFinished {
+	Yes,
+	No,
+}
 
-		if version < 9 {
-			v9::migrate::<T>(&mut weight);
-		}
+/// A trait that allows to migrate storage from one version to another.
+///
+/// The migration is done in steps. The migration is finished when
+/// `step()` returns `IsFinished::Yes`.
+pub trait Migrate: Codec + MaxEncodedLen + Default {
+	/// Returns the version of the migration.
+	const VERSION: u16;
 
-		StorageVersion::new(9).put::<Pallet<T>>();
-		weight.saturating_accrue(T::DbWeight::get().writes(1));
+	/// Returns the maximum weight that can be consumed in a single step.
+	fn max_step_weight() -> Weight;
 
-		weight
-	}
+	/// Process one step of the migration.
+	///
+	/// Returns whether the migration is finished and the weight consumed.
+	fn step(&mut self) -> (IsFinished, Weight);
```

Comment:
It would be helpful to add a warning here to aware developers that `step` should be small enough to fit in a single block execution, otherwise the entire pallet would get blocked

Reply:
agreed, but you think about this this is true for everything we do in the runtime, any kind of weight that we produce should be lower than T::BlockWeights::get().max_block;

Reply:
add that to the integrity_test 
[453489f0fc37706719a0ee920bfe19fdd846e96d](https://github.com/paritytech/substrate/pull/14045/commits/453489f0fc37706719a0ee920bfe19fdd846e96d)